### DISCUSSION TITLE
type_annotation_map: possible to map all List and Dict type to SQLAlchemy JSON?

### DISCUSSION THREAD

**Discussion by abceleung** [sqlalchemy/sqlalchemy#11066] [Usage Questions] [answered]
Currently I am doing this:
```python
from typing import Dict, List, Any
from sqlalchemy import JSON
from sqlalchemy.orm import Mapped, mapped_column

reg = registry(
    type_annotation_map={
        Dict[str, str]: JSON,
        List[str]: JSON,
        List[
            Dict[
                str,
                str
                | Dict[
                    str, str | Dict[str, str | Dict[str, Dict[str, Any]]]
                ],
            ]
        ]: JSON,
        List[
            Dict[str, str | Dict[str, str]]
            | Dict[
                str,
                str
                | Dict[
                    str, str | List[Dict[str, str | int | Dict[str, str]]]
                ],
            ]
        ]: JSON,
        Dict[
            str, str | Dict[str, List[Dict[str, str | Dict[str, str]]]]
        ]: JSON,
        Dict[str, int]: JSON,
    }
)

class ModelBase(DeclarativeBase, MappedAsDataclass):
    registry = reg

class Thread(ModelBase):
    __tablename__ = "thread"

    thread_id: Mapped[int] = mapped_column(
        init=False, primary_key=True, autoincrement=True
    )
    id: Mapped[str] = mapped_column(String(50), unique=True)
    object: Mapped[str] = mapped_column(String(10))
    created_at: Mapped[int] = mapped_column()
    metadata_: Mapped[Dict[str, str]] = mapped_column("metadata")

```
As you can see that it is a mess. 

I tried rewrite `reg` to this:
```python
from typing import Dict, List, Any
from sqlalchemy import JSON

reg = registry(
    type_annotation_map={
        Dict[str,Any]: JSON,
        List[Any]: JSON
    }
)
```
But SQLAlchemy threw error : `Could not locate SQLAlchemy Core type for Python type typing.Dict[str, str] inside the 'metadata' attribute Mapped annotation`.
Is there a way to let SQLAlchemy map all Dict and List type to JSON?

**Comment by zzzeek** [comment, marked as answer]
> Currently I am doing this:
> 
> ```python
> from typing import Dict, List, Any
> from sqlalchemy import JSON
> from sqlalchemy.orm import Mapped, mapped_column
> 
> reg = registry(
>     type_annotation_map={
>         ty.Dict[str, str]: JSON,
>         ty.List[str]: JSON,
>         ty.List[
>             ty.Dict[
>                 str,
>                 ty.Dict[str, str]
>                 | ty.Dict[str, str | ty.Dict[str, str | ty.Dict[str, ty.Any]]],
>             ]
>         ]: JSON,
>         ty.List[
>             ty.Dict[str, str | ty.Dict[str, str]]
>             | ty.Dict[
>                 str,
>                 str
>                 | ty.Dict[
>                     str, str | ty.List[ty.Dict[str, str | int | ty.Dict[str, str]]]
>                 ],
>             ]
>         ]: JSON,
>         ty.Dict[
>             str, str | ty.Dict[str, ty.List[ty.Dict[str, str | ty.Dict[str, str]]]]
>         ]: JSON,
>         ty.Dict[str, int]: JSON,
>     }
> )
> 
> class ModelBase(DeclarativeBase, MappedAsDataclass):
>     registry = reg
> 
> class Thread(ModelBase):
>     __tablename__ = "thread"
> 
>     thread_id: Mapped[int] = mapped_column(
>         init=False, primary_key=True, autoincrement=True
>     )
>     id: Mapped[str] = mapped_column(String(50), unique=True)
>     object: Mapped[str] = mapped_column(String(10))
>     created_at: Mapped[int] = mapped_column()
>     metadata_: Mapped[ty.Dict[str, str]] = mapped_column("metadata")
> ```
> 
> As you can see that it is a mess.

sure, so automate it with code

```
def _json_dict(level=0):
    yield int
    yield str
    if level >2:
        return

    for subtype in _json_dict(level+1):
        yield subtype
        yield ty.Dict[str, subtype]
        yield ty.List[subtype]

reg = registry(
    type_annotation_map={
        elem: JSON for elem in _json_dict()
    }
)

```

> But SQLAlchemy threw error : `Could not locate SQLAlchemy Core type for Python type typing.Dict[str, str] inside the 'metadata' attribute Mapped annotation`. 

that's because the parameters to generic types aren't expanded out or anything like that.


> Is there a way to let SQLAlchemy map all Dict and List type to JSON?

use a recipe similar to the one above

**Reply by abceleung** [reply]
Thanks for the example code, though I have to modify it like this (and it takes a while to run):

```python
import typing as ty

from sqlalchemy import JSON


def _json_dict(level: int = 0):
    yield int

    yield str

    if level > 7:
        yield ty.Any

    if level > 8:
        return

    for subtype in _json_dict(level + 1):
        yield subtype

        yield ty.Dict[str, subtype]

        yield ty.List[subtype]

        if subtype != str:
            yield ty.Union[str, subtype]

# This line takes 9 seconds to run, lol
typeDict = {elem: JSON for elem in _json_dict()}
```

Then typeDict would contain my most nested type:
```python
typeDict[
    ty.List[
        ty.Dict[
            str,
            str
            | ty.Dict[
                str, str | ty.Dict[str, str | ty.Dict[str, ty.Dict[str, ty.Any]]]
            ],
        ]
    ]
]
# Outputs sqlalchemy.sql.sqltypes.JSON
```

Will spend some time to optimize this.
